torch/utils/data/dataloader.py

```diff
@@ -21,29 +24,52 @@
 default_collate = _utils.collate.default_collate
 
 
+class _DataLoaderStrategy(object):
+    Map = 0
+    MapWithBatchedRead = 1
+    Iterable = 2
+
+
 class DataLoader(object):
     r"""
     Data loader. Combines a dataset and a sampler, and provides
     single- or multi-process iterators over the dataset.
 
+    The :class:`~torch.utils.data.DataLoader` applies one of the following
+    data loading strategies based on the constructor arguments:
+
+    * batched loading from a map-style dataset (default)
+
+    * Loading individual members of a map-style dataset
+
+    * loading from an iterable-style dataset
+
+
+    See `Dataset Types`_ and `Data Loading Strategies`_ for more details.
+
     Arguments:
         dataset (Dataset): dataset from which to load the data.
         batch_size (int, optional): how many samples per batch to load
             (default: ``1``).
         shuffle (bool, optional): set to ``True`` to have the data reshuffled
             at every epoch (default: ``False``).
         sampler (Sampler, optional): defines the strategy to draw samples from
-            the dataset. If specified, ``shuffle`` must be False.
-        batch_sampler (Sampler, optional): like sampler, but returns a batch of
+            the dataset. If specified, :attr:`shuffle` must be False.
+        batch_sampler (Sampler, optional): like :attr:`sampler`, but returns a batch of
             indices at a time. Mutually exclusive with :attr:`batch_size`,
             :attr:`shuffle`, :attr:`sampler`, and :attr:`drop_last`.
         num_workers (int, optional): how many subprocesses to use for data
-            loading. 0 means that the data will be loaded in the main process.
+            loading. ``0`` means that the data will be loaded in the main process.
             (default: ``0``)
-        collate_fn (callable, optional): merges a list of samples to form a mini-batch.
-        pin_memory (bool, optional): If ``True``, the data loader will copy tensors
+        collate_fn (callable, optional): merges a list of samples to form a
+            mini-batch of Tensor(s).  Used when using batched loading from a
+            map-style dataset.
+        convert_fn (callable, optional): converts a sample to Tensor(s).  Used
```

Comment:
Could you elaborate a bit in terms of how and when collate_fn and convert_fn are used? I'm a little confused in terms of how the two differ.

Reply:
`collate_fn` is used when batch loading (i.e., the current supported kind), to collate a list of samples into a batched sample, e.g., the two element batch `[(A1, A2), (B1, B2)]` into `torch.as_tensor([A1, B1]), torch.as_tensor([A2, B2])`.

`convert_fn` just does conversion, so if a sample is `[(A, B), (C, D)]` it can just call `as_tensor` on each.

Reply:
I see, could you also elaborate a bit in the comments as well so the difference is clear?

Reply:
Yes, I will do it in the doc.